### PR TITLE
vendor: buildkit v0.8.0-rc2, docker, containerd, and dependencies

### PR DESCRIPTION
**Pull request by thaJeztah**
### vendor: golang.org/x/crypto c1f2f97bffc9c53fc40a1a28a5b460094c0050d9

(included from https://github.com/moby/moby/pull/41690)

full diff: https://github.com/golang/crypto/compare/75b288015ac94e66e3d6715fb68a9b41bf046ec2...c1f2f97bffc9c53fc40a1a28a5b460094c0050d9

relevant changes:

- pkcs12: document that we use the wrong PEM type
- pkcs12: drop PKCS#12 attributes with unknown OIDs
- ocsp: Improve documentation for ParseResponse and ParseResponseForCert
- ssh: improve error message for KeyboardInteractiveChallenge
- ssh: remove slow unnecessary diffie-hellman-group-exchange primality check

other changes (not in vendor);

- ssh/terminal: replace with a golang.org/x/term wrapper
    - Deprecates ssh/terminal in favor of golang.org/x/term
- ssh/terminal: add support for zos
- ssh/terminal: bump x/term dependency to fix js/nacl
- nacl/auth: use Size instead of KeySize for Sum output
- sha3: remove go:nocheckptr annotation

### vendor: github.com/tonistiigi/fsutil 0834f99b7b85462efb69b4f571a4fa3ca7da5ac9

full diff: https://github.com/tonistiigi/fsutil/compare/ae3a8d753069d0f76fbee396457e8b6cfd7cb8c3...0834f99b7b85462efb69b4f571a4fa3ca7da5ac9

- walker: fix notadir error
- improving error returns
    - more typed errors
    - remove extra verbosity (eg. PathError already contains action and path)
    - ensure stack traces are added to errors
- various testing and linting fixes
- copy: use Clonefileat from golang.org/x/sys/unix on macOS
- go.mod: update opencontainers/go-digest v1.0.0
- github: test go1.15

### vendor: github.com/moby/term bea5bbe245bf407372d477f1361d2ff042d2f556

full diff: https://github.com/moby/term/compare/7f0af18e79f2784809e9cef63d0df5aa2c79d76e...bea5bbe245bf407372d477f1361d2ff042d2f556

- Fix windows integer overflow on GOOS=windows, GOARCH=arm
- go.mod: github.com/creack/pty v1.1.11
  - v1.1.11: Add arm support for OpenBSD
  - v1.1.10: Fix CTTY to work with go1.15
- CI: fix Go version matrix, and drop go 1.12, add go 1.15
- CI: remove "sudo" to fix incorrect Go versions (incorrect PATH, GOROOT)

### vendor: update docker, buildkit, containerd

diffs:

- full diff: https://github.com/docker/docker/compare/af34b94a78a194ae9bce48a200241d08c05b7187...6c0a036dce2051cc0d73221965e76e3d2e6a8a3a
- full diff: https://github.com/moby/buildkit/compare/4d1f260e8490ec438ab66e08bb105577aca0ce06...v0.8.0-rc2
- full diff: https://github.com/containerd/containerd/compare/v1.4.1...0edc412565dcc6e3d6125ff9e4b009ad4b89c638

New dependencies:

- go.opencensus.io v0.22.3
- github.com/containerd/typeurl v1.0.1
- github.com/golang/groupcache 869f871628b6baa9cfbc11732cdf6546b17c1298

### CODE REVIEW

**Comment by thaJeztah** [comment]
We need to look if we can refactor containerd and/or buildkit, as this is now pulling in a _ton_ of new dependencies:

Screenshot 2020-11-17 at 14 23 33

**Comment by thaJeztah** [comment]
Trying to map why/where the new dependencies are used;

github.com/containerd/fifo
    -> imported in github.com/containerd/containerd/cio/io_unix.go
        -> imported in github.com/containerd/containerd (root)

github.com/containerd/typeurl
    -> used in github.com/containerd/containerd (root)
    -> imported in buildkit

github.com/containerd/ttrpc
    -> imported in github.com/containerd/containerd/plugin/plugin.go
        -> part of `TTRPCService` interface: must this be in the "plugin" package, or could this be separate?

github.com/google/uuid
    -> imported in github.com/containerd/containerd/services/introspection/local.go
        -> used in github.com/containerd/containerd (root);
            -> imported in github.com/containerd/containerd/services.go
                -> used in containerd.WithIntrospectionService()
            -> imported in github.com/containerd/containerd/client.go
                -> used in client.IntrospectionService()

github.com/opencontainers/selinux
    -> used in github.com/containerd/containerd (root):
        -> imported in github.com/containerd/containerd/container.go
            -> used in container.NewTask()
            -> used for a single function (`label.FormatMountLabel`)

github.com/syndtr/gocapability
    -> used in github.com/containerd/containerd/oci
        -> imported in github.com/containerd/containerd/oci/spec_opts.go
            -> used in a single function: GetAllCapabilities()
            -> could this be moved to a separate package? github.com/containerd/containerd/oci/capabilities (?)

go.opencensus.io
    -> used in github.com/Microsoft/hcsshim/internal/ and github.com/Microsoft/hcsshim/internal/hcs/
        -> imported in github.com/containerd/containerd/mount/mount_windows.go
        -> imported in github.com/containerd/containerd/archive/tar_windows.go
        -> imported in github.com/containerd/containerd/sys/filesys_windows.go

github.com/golang/groupcache/lru
    -> used in go.opencensus.io/trace/lrumap.go

**Comment by thaJeztah** [comment]
opened https://github.com/moby/buildkit/pull/1808 to fix the dependencies, and pushed a temporary commit that vendors buildkit from my fork

File: vendor.conf [hunk 1]
```diff
@@ -67,12 +69,13 @@ github.com/sirupsen/logrus                          6699a89a232f3db797f2e2806398
 github.com/spf13/cobra                              86f8bfd7fef868a174e1b606783bd7f5c82ddf8f # v1.1.1
 github.com/spf13/pflag                              2e9d26c8c37aae03e3f9d4e90b7116f5accb7cab # v1.0.5
 github.com/theupdateframework/notary                d6e1431feb32348e0650bf7551ac5cffd01d857b # v0.6.1
-github.com/tonistiigi/fsutil                        ae3a8d753069d0f76fbee396457e8b6cfd7cb8c3
+github.com/tonistiigi/fsutil                        c3ed55f3b48161fd3dc42c17ba09e12ac52d57dc
```
Comment (thaJeztah) [inline vendor.conf:72]: note to self: If we use this version of buildkit, also need to update this to current master (to match)

File: vendor/github.com/moby/buildkit/session/auth/authprovider/authprovider.go [hunk 2]
```diff
@@ -28,17 +50,80 @@ type authProvider struct {
 	mu sync.Mutex
 }
 
+func (ap *authProvider) SetLogger(l progresswriter.Logger) {
```
Comment (thaJeztah) [inline vendor/github.com/moby/buildkit/session/auth/authprovider/authprovider.go:53]: Need to connect this during build

Comment (thaJeztah) [reply vendor/github.com/moby/buildkit/session/auth/authprovider/authprovider.go:53]: @tonistiigi  I had a look, but not sure how to handle this; the authProvider is added here; https://github.com/docker/cli/blob/88c6089300a82d3373892adf6845a4fed1a4ba8d/cli/command/image/build_buildkit.go#L179

To call `SetLogger`, looks like some ugly "interface" matching is needed, similar to https://github.com/moby/buildkit/blob/4177e7ca28c9918070a5ef30d4fa8a43d911e08d/cmd/buildctl/build.go#L269-L275

(curious; should `authProvider` be exported, or a `SetLogger()` interface be defined somewhere?)

```go
if s, ok := at.(interface {
	SetLogger(progresswriter.Logger)
}); ok {
	w := mw.WithPrefix("", false)
	s.SetLogger(func(s *client.SolveStatus) {
		w.Status() <- s
	})
```

However in the CLI, I don't think we have a `progresswriter.Logger` at this point in the code, as those are created in `doBuild` https://github.com/docker/cli/blob/88c6089300a82d3373892adf6845a4fed1a4ba8d/cli/command/image/build_buildkit.go#L276-L286


Perhaps I misunderstood how the logger should be set (can it be done at a later stage? if so, how do I get a reference to the `authProvider` later on?

**Review by tonistiigi — APPROVED** [review]

**Review by tiborvass — APPROVED** [review]

**Pull request merged**